podman_hpc/podman_hpc.py

```diff
@@ -164,6 +165,72 @@ def add_args(parser, confs):
                             help=v.get("help"))
 
 
+def filter_podman_subcommand(podman_bin, subcommand, podman_args):
+    """ Filter invalid arguments from an argument list 
+    for a given podman subcommand based on its --help text. """
+    # extract valid flags from subcommand help text, and populate an arg parser
+    opt_regex = re.compile("^\s*(?:(-\w), )?(--\w[\w\-]+)(?:\s(\w+))?")
+    p = argparse.ArgumentParser(exit_on_error=False)
+    with os.popen(' '.join([podman_bin,subcommand,'--help'])) as f:
+        for line in f:
+            opt = opt_regex.match(line)
+            if opt:
+                action = 'store' if opt.groups()[2] else 'store_true'
+                flags = [flag for flag in opt.groups()[:-1] if flag]
+                p.add_argument(*flags,action=action)
+    # remove unknown args from the podman_args
+    subcmd_args = podman_args.copy()
+    unknowns = p.parse_known_args(subcmd_args)[1]
+    uk_safe={} # indices of valid args that string== unknowns
+    while unknowns:
+        ukd={} # candidate indices of where to remove unknowns
+        for uk in set(unknowns):
+            ukd[uk]=[idx for idx,arg in enumerate(subcmd_args) if (arg==uk and idx not in uk_safe.get(uk,[]))]
+        uk=unknowns.pop(0) 
+        # find and remove an invalid occurence of uk
+        while True:
+            args_tmp = subcmd_args.copy()
+            args_tmp.pop(ukd[uk][0])
+            try:
+                if p.parse_known_args(args_tmp)[1]==unknowns:
+                    subcmd_args.pop(ukd[uk][0])
+                    break
+            except argparse.ArgumentError:
+                pass
+            uk_safe.setdefault(uk,[]).append(ukd[uk].pop(0))
+    return [podman_bin,subcommand]+subcmd_args
+
+
+def shared_run_args(podman_args,image,container_name='hpc'):
+    """ Construct argument list for `podman run` and `podman exec`
+    by filtering flags passed to `podman shared-run` """
+    print("calling shared_run_args with podman_args:")
+    print(f"\t{podman_args}")
+    
+    # generate valid subcommands from the given podman_args
+    prun =filter_podman_subcommand(podman_args[0],'run',podman_args)
+    pexec=filter_podman_subcommand(podman_args[0],'exec',podman_args)
+
+    prun[2:2] = ['--rm','-d','-v','/global/homes/d/dfulton/repos/podman-hpc/bin/exec-wait:/bin/exec-wait','--name',container_name]
```

Comment:
i think you probably want to remove this ref, or move it into the podman-hpc project

Reply:
Good catch, this should be pulled in with a custom hook, so replaced it with `--exec-wait` flag

Reply:
Just as a note, this gets replaced with the `click` refactor anyhow.